plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -0,0 +1,274 @@
+
+    This stage creates a ProgressBar named 'Downloading Artifacts' that counts the number of
+    downloads completed. Since it's a stream the total count isn't known until it's finished.
+
+    This stage drains all available items from `in_q` and starts as many downloaders as possible.
+
+    Args:
+        max_concurrent_downloads (int): The maximum number of concurrent downloads this stage will
+            run. Default is 100.
+
+    Returns:
+        An object containing the artifact_downloader stage to be included in a pipeline.
+    """
+
+    def __init__(self, max_concurrent_downloads=100):
+        self.max_concurrent_downloads = max_concurrent_downloads
+
+    async def stage(self, in_q, out_q):
+        """
+        Download undownloaded Artifacts, but don't save them in the db.
+
+        Args:
+            in_q (:class:`asyncio.Queue`): The queue to receive
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from that may have
+                undownloaded files.
+            out_q (:class:`asyncio.Queue`): The queue to put
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects into, all of which have
+                files downloaded.
+
+        """
+        pending = set()
+        incoming_content = []
+        outstanding_downloads = 0
+        shutdown = False
+        saturated = False
+        with ProgressBar(message='Downloading Artifacts') as pb:
+            while True:
+                if not saturated:
+                    try:
+                        content = in_q.get_nowait()
+                    except asyncio.QueueEmpty:
+                        if not incoming_content and not shutdown and not pending:
+                            content = await in_q.get()
+                            incoming_content.append(content)
+                            continue
+                    else:
+                        incoming_content.append(content)
+                        continue
+
+                for i, content in enumerate(incoming_content):
+                    if content is None:
+                        shutdown = True
+                        continue
+                    downloaders_for_content = []
+                    for declarative_artifact in content.d_artifacts:
+                        if declarative_artifact.artifact._state.adding:
+                            # this needs to be downloaded
+                            downloader = declarative_artifact.remote.get_downloader(
+                                declarative_artifact.url
+                            )
+                            next_future = asyncio.ensure_future(downloader.run())
+                            downloaders_for_content.append(next_future)
+                    if not downloaders_for_content:
+                        await out_q.put(content)
+                        continue
+
+                    async def return_content_for_downloader(c):
+                        return c
```

Comment:
I don't see the point of using this closure since it will immediately return `c`.  Can you help me understand this?

Reply:
Yes, it looks a little strange but it makes the code elagent. This closure has no purpose except to maintain a reference to c and then return it so that the closure completes. That reference is then returned in the list of returned results as part of `gather(...)`. This allows the content unit and specifically its artifacts to all come back together after all the artifacts are downloaded. That reference to the content unit is stripped out of the results [here](https://github.com/pulp/pulp/pull/3559/files#diff-026c51b6860b1ee72f80e464a00a2228R190)